### ISSUE TITLE
Build without Bluebird?

### ISSUE THREAD

**Issue by seansd** [FormidableLabs/rapscallion#115]
Bluebird is a promise library, and most of that work is for client-side functionality.  I'm wanting to use your package server-side only, and we already are on Node 6+ which supports promises, not too mention we have our own promise shim.. . any chance of removing that dependency. . . ?

**Comment by divmain** [comment]
Can you help me understand your use case?  As far as I know, Bluebird still performs better than V8 native Promises, so there's a compelling reason to keep it as a dependency.  Bluebird shouldn't make it onto your client-side bundle, so if you're seeing something like that, please let me know.

**Comment by seansd** [comment]
Simply a package dependency deal.  I'd rather not include something that's not actually needed.  We use native promises already with great success and while bluebird might perform better today, eventually that won't be the case (in fact we'll probably be switching to Node8 pretty soon).  I certainly can fork the project but that seemed a bit heavy handed.

**Comment by ryan-roemer** [comment]
@seansd -- Let me see if I can summarize the use case / problem area here (and please correct me if I'm wrong!):

- You use `rapscallion` only on the server-side in a node6+ environment with native `Promise` support.
- Thus, you technically don't need `bluebird` as a dependency.
- `rapscallion` has a prod `dependencies` on `bluebird`. `bluebird` has no `dependencies`. Using `npm pack bluebird` to see the size that means when you take a dependency on `rapscallion` you are technically getting an extra `132KB` of downloads from npm that you don't need.

... is that all correct?

**Comment by divmain** [comment]
Hi @seansd, this will probably have to be deferred until Node actually catches up performance-wise.  You're free to use native promises in your own code, though - Promises implement a specific interface and should all be compatible, so long as they're written to spec.

Here's where things lie right now.  This is a "real world" benchmark taken from Bluebird's repository and run on Node v8.1.2.

| file | time(ms) | memory(MB) | speedup |
| ---- | -------- | ---------- | ------- |
| callbacks-baseline.js | 132 | 31.37 | 153% |
| callbacks-suguru03-neo-async-waterfall.js | 181 | 39.7 | 121% |
| promises-bluebird-generator.js | 243 | 41.95 | 115% |
| promises-bluebird.js | 270 | 48.08 | 100% |
| promises-cujojs-when.js | 388 | 63.08 | 76% |
| promises-tildeio-rsvp.js | 469 | 77.62 | 62% |
| promises-then-promise.js | 481 | 75.18 | 64% |
| promises-lvivski-davy.js | 517 | 100.47 | 48% |
| promises-ecmascript6-native.js | 608 | 99.29 | 48% |

```
Platform info:
Darwin 16.7.0 x64
Node.JS 8.1.2
V8 5.8.283.41
Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz × 8
```

As you can see, using native promises incurs a roughly 2x memory cost and a 2x time cost.  That's an unacceptable regression, and would likely undermine the utility of Rapscallion as a useful tool.

Native promises are definitely on the table, but only once they catch up to some of the faster JS implementations.

**Comment by seansd** [comment]
@ryan-roemer Yes, and also not just the extra code, but also the overhead of yet another dependency to manage when building a large site.  (Thankfully bluebird itself does not have any dependencies on it's own).

Seems like it could at least be an option to allow for support for native promises. . . but I understand your point about the performace (jeesh maybe i'll just forbid promises all together when simple callbacks are still the winner :) ).  I'll confer with my team but in the mean time it's a suggestion.

Thank for for all the information and being responsive.
